Added OpenTracing Tags to diagnostic Activities

RDalziel opened the pull request:
Added message types to logs as currently it can be onerous to crawl logs without it. My experience is with app insights here - happy to be corrected if this doesn't fit activity naming guidelines. It's just pretty horrible in app insights at least.

![image](https://user-images.githubusercontent.com/3485935/70947731-b3d8ef80-2051-11ea-962f-c94c5ef0b4da.png)


<!--
Thank you for sending the PR!

If you changed any code, please provide us with clear instructions on how you verified your changes work. Bonus points for screenshots!

Happy contributing!
-->

phatboyg commented:
So, help me understand what this adds and can you provide some links to how it is done with other libraries around operation name for the activity?

RDalziel commented:
Unfortunately i don't have anything for you there @phatboyg 

This change relates to trying to achieve similar results to what was available in 5.x. When parsing logs on the new setup - the log messages show as in the screenshot in the original post.

Then you have a more complex setup with a state machine firing out several commands - it can be a bit onerous to parse the logs to find the particular path you're interested in.

RDalziel commented:
Looking through https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.DiagnosticSource/src/ActivityUserGuide.md

It seems to me the question here is "How would we want to group and filter the logs?"

For me - it's useful to filter/group by "Consume CreateUserCommand" rather than just "Consume"

RDalziel commented:
For a better view - see how ASB Diasnotic logs look alongside MT :
![image](https://user-images.githubusercontent.com/3485935/71017635-bf2e2880-20ee-11ea-9f5d-bab08e2779fb.png)

RDalziel commented:
@NooNameR @maldworth What's your thoughts here?

NooNameR commented:
probably it isn't huge change, so it could make sense in some situations

phatboyg commented:
What statement is creating the left heading:

mtdiagnost Messages/Zing

I still don't see anywhere in other uses where they are creating operation names based upon URL's or types or anything like it -- and it feels wrong to me.

RDalziel commented:
That's Microsoft.Azure.ServiceBus.Core generating those - just found the reference here:

https://github.com/Azure/azure-service-bus-dotnet/blob/e1f6e531ac8dc2417d59c03a983d9987e3850989/src/Microsoft.Azure.ServiceBus/ServiceBusDiagnosticsSource.cs

At a quick glance i can't see where the first log is in here. The above makes me think it's closer to your thinking, Chris, Where they don't include entity information in the activity name.

RDalziel commented:
From a filtering aspect though - look how MT looks next to ASB logs - 
![image](https://user-images.githubusercontent.com/3485935/71100166-6d4bd800-21ac-11ea-9703-8c30465e5e1b.png)

![image](https://user-images.githubusercontent.com/3485935/71100242-94a2a500-21ac-11ea-95ba-0215f127e33e.png)

phatboyg commented:
In that library, it is passing an anonymous object to the Activity.Start method, and in many cases that's been suggested as using an IDictionary<string,object>, so that it can be used by other libraries to extract information. Feels like the approach has merit, but this is one example. I'd like to see how ASP.NET Core is doing it next.

RDalziel commented:
All i can find on ASP.NET is the following - https://github.com/aspnet/AspNetCore/blob/5b2f3fb5f7f24ac3e91c5150a55cc411b2b36b76/src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs

phatboyg commented:
Good detail. And it looks like they pass the httpContext to activity start, which is likely used by the AppInsights listener to extract the URL and post it back. Unless App Insights knows how to parse the types passed by masstransit, it won't be able to use that information. ANd this is always going to be listener-specific. I think that's why the string/object dictionary makes so much sense as arguments to start/stop activity.

RDalziel commented:
So if i'm understanding this right - we should bin this change and move towards looking at how Application Insights translates diagnostic logs? Perhaps create an App specific listener to cover it?

RDalziel commented:
An equivalent to this - 
https://github.com/microsoft/ApplicationInsights-dotnet-server/blob/develop/WEB/Src/DependencyCollector/Shared/TelemetryDiagnosticSourceListener.cs

NooNameR commented:
I'd prefer to use custom tags:
https://github.com/opentracing/specification/blob/master/semantic_conventions.md

RDalziel commented:
You can see that App Insights seem to be making use of those conventions.

NooNameR commented:
@RDalziel that's why I'd like to use them, if it is OpenTracing's convention so every provider should use them

RDalziel commented:
Not sure where Message Type falls into those conventions, though.

RDalziel commented:
What's the state of OpenTelemetry vs OpenTracing??

https://github.com/microsoft/ApplicationInsights-Home/issues/391#issuecomment-528679853

maldworth commented:
Looks like OpenTelemetry is the way to go there. So the question is, use Open Tracing semantics in the interim (maybe a specific listener as @RDalziel said?) until OpenTelemetry is ready for prime time with App Insights?

NooNameR commented:
@maldworth I think we could use convention. I'm pretty sure that Open Tracing and Open Telemetry use the same naming, to make it easy for developers to switch between

RDalziel commented:
Hm. Yes and no @NooNameR 
OpenTelemetry and OpenTracing seem to vary on networking, Which seem like some of the useful tags for messaging currently (networkig).

There's a PR here for some Messaging tags on OpenTelemetry but still very much looks in discussion.

How about i have a go at the tags and see where i can so we at least have a baseline?

RDalziel commented:
@NooNameR @maldworth @phatboyg 

Added a few OpenTracing tags - supported in App Insights without any additional work. I've left my sample app in there for now, Will take it out afterwards.

OpenTelemetry doesn't appear to have 'peer.host' & 'peer.address' but may have more support for asynchronous message heads soon - 
https://github.com/open-telemetry/opentelemetry-specification/pull/395

maldworth commented:
This is great, thank for your work on this. I always think having more samples isn't a bad thing. You can reference them in the docs as well. But perhaps moving them to a separate repo, maybe **Sample-Telemetry**, or **Sample-Diagnostics**. The repo could start with maybe App Insights (and maybe Jaeger?), eventually also have example of OpenTelemetry when it's ready for prime time?

RDalziel commented:
Agreed & cheers @maldworth  I've cleaned up the branch and removed the sample. Was intending to do additional work looking at docs + the issue of nesting the activities from root node to the final consumer/send.

NooNameR commented on src/MassTransit/DiagnosticSourceExtensions.cs:
```diff
@@ -24,6 +24,9 @@ public static void AddSendContextHeaders(this StartedActivity? startedActivity,
             activity.AddTag(DiagnosticHeaders.InitiatorId, context.InitiatorId);
             activity.AddTag(DiagnosticHeaders.SourceAddress, context.SourceAddress);
             activity.AddTag(DiagnosticHeaders.DestinationAddress, context.DestinationAddress);
+            activity.AddTag("span.kind", "producer");
```
could you please move it to DiagnosticHeaders?
also we have our own so we could change them to use right convention, for eg: `DiagnosticHeaders.DestinationAddress`.

RDalziel replied:
Updated - i've not updated existing headers though. Unsure what's currently using them and unsure the naming doesn't completely align. What you think on the naming i've gave?

NooNameR replied:
"span.kind" -> DiagnosticHeaders.ServiceKind
DiagnosticHeaders.DestinationAddress -> "peer.address"

RDalziel replied:
Updated

NooNameR commented:
@phatboyg what do you think about existing headers? should we change them? I'd like to adapt them to use right convention

phatboyg commented:
Looks like it's getting there. Shouldn't we have Kind of Consumer, Saga, Activity, etc. for all the things?

NooNameR commented:
I'm not sure but Saga/Activity isn't consumer by the nature? but, tbh it is a good question...

RDalziel commented:
Sounds like it'd be going against the conventions. They only have definitions for Client, Server, Producer & Consumer.

RDalziel commented:
@NooNameR @phatboyg @NooNameR Added for the remaining areas - not entirely sure the correct metadata to show in each scenario but hopefully we can iron that out on comments :)

NooNameR commented:
seems ok for me. do you have any working sample, or could you make a screenshot how it looks?

RDalziel commented:
@NooNameR Just the one above for App Insights - there may be some work on continued work for some of the other providers. Just convers PubSub on Consumers. I checked the rest on existing unit tests.

![image](https://user-images.githubusercontent.com/3485935/71562438-1c739380-2a78-11ea-9af0-57ab217dd65d.png)

phatboyg commented on src/MassTransit/DiagnosticSourceExtensions.cs:
```diff
@@ -25,9 +25,8 @@ public static void AddSendContextHeaders(this StartedActivity? startedActivity,
-            activity.AddTag(DiagnosticHeaders.PeerHost, context.DestinationAddress.Host);
-            activity.AddTag(DiagnosticHeaders.PeerAddress, context.DestinationAddress.AbsolutePath);
+            activity.AddTag(DiagnosticHeaders.ServiceKind, DiagnosticHeaders.Kind.Producer);
+            activity.AddTag(DiagnosticHeaders.DestinationHost, context.DestinationAddress.Host);
```
This is worthless, it's just the message broker host name.

RDalziel replied:
I don't have a particular use for it - really just trying to make use of peer.host. I guess i could see some scenarios where someone may be working with different message broker instances and wants to isolate logs in that way.

phatboyg commented on src/MassTransit/Pipeline/Filters/ConsumerMessageFilter.cs:
```diff
@@ -48,8 +48,8 @@ async Task IFilter<ConsumeContext<TMessage>>.Send(ConsumeContext<TMessage> conte
-            activity?.AddTag(DiagnosticHeaders.Kind, DiagnosticHeaders.SpanKind.Consumer);
-            activity?.AddTag(DiagnosticHeaders.PeerAddress, TypeMetadataCache<TConsumer>.ShortName);
+            activity?.AddTag(DiagnosticHeaders.ServiceKind, DiagnosticHeaders.Kind.Consumer);
+            activity?.AddTag(DiagnosticHeaders.DestinationAddress, TypeMetadataCache<TConsumer>.ShortName);
```
How is the consumer type a peer.address? Seems wrong, should be the input queue address.

RDalziel replied:
My thinking was to distinguish between multiple consumers that might be on the same receive endpoint.

RDalziel replied:
Note also that the Receive traces will already have the input address

phatboyg commented:
Seems like there are some tags that make sense to use: https://github.com/opentracing/specification/blob/master/semantic_conventions.md

RDalziel commented:
I've stayed away from "message_bus.destination" for now as it's not represented in the OpenTelemetry spec which will supersede OpenTracing.

https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-span-general.md

You can see that peer names are also updating there so there will probably be continued work here as OpenTelemetry becomes more concrete.

I've noticed the peer.host should be peer.hostname from current open tracing and usage here - https://github.com/microsoft/ApplicationInsights-dotnet-server/blob/develop/WEB/Src/DependencyCollector/Shared/TelemetryDiagnosticSourceListener.cs 
Will update this now!

phatboyg commented:
It's helpful, I'm trying to understand how it all works. And I have a separate branch that includes a rework of the startActivity method which includes overloads so that the dictionary (key,object) is properly passed as the StartActivity argument, and that the tags are added before calling Start() on the activity. Should dry up the code a bit as well.

The nested Receive/Consume context makes sense, since Consume is just ends up being a span (well, an activity) within the receive. If that's the case, the input address should be baggage, vs a tag, so that it is copied to the consume activity.

THe real goal is so that this all renders properly in the tools (including but not limited to App Insights).

phatboyg commented:
This is with RabbitMQ, using Application Insights...
![image](https://user-images.githubusercontent.com/20041/71649742-bd5f8a00-2cd6-11ea-9e04-67a6e5e45cf1.png)

phatboyg commented:
Now, with a separate saga added, to show the separate receive paths. I'm pretty close to pushing this up to develop along with documentation on how to setup Application Insights.

![image](https://user-images.githubusercontent.com/20041/71674180-30e9b180-2d40-11ea-83c7-41276875d81d.png)

RDalziel commented:
Awesome stuff! How about other dependency logs ? Do they best underneath too ? For example, if you do a http request in a consumer...is it connected to the same tree ?

phatboyg commented:
Also, the detail on a saga trace is pretty useful:

![image](https://user-images.githubusercontent.com/20041/71674402-b0778080-2d40-11ea-83da-0e3eea418434.png)

phatboyg commented:
Okay, so much easier on the eyes!

![image](https://user-images.githubusercontent.com/20041/71677127-9beab680-2d47-11ea-9742-2123e6f42142.png)

RDalziel commented:
App insights UI is not the best at dealing with longer op names...but makes sense to trim where possible...great stuff!

phatboyg commented:
Okay, now with a routing slip... (yeah, I'm sorta having fun getting this all tied together).

![image](https://user-images.githubusercontent.com/20041/71679414-28e43e80-2d4d-11ea-9404-a32e82390fb4.png)

phatboyg commented:
See #1657
